BEC-290: Add needed Airnode options to Airkeeper config

acenolaza opened the pull request:
Closes #69 

I have to admit that I'm not totally convinced this is better that what we currently have. Yes, we now have a single config file but the code now has a bunch of things (validator mostly) that are never going to be used and will probably make more difficult to keep the Airnode config.json values in sync. I'd like your opinion on this tho.

I've created this PR as a draft because I need to fix some tests and update the README

Ashar2shahid commented on config/airkeeper.example.json:
```diff
@@ -3,14 +3,71 @@
   "airnodeXpub": "xpub6CjvSJ3sybHuVaYnQsCvNQnXfNrMusXEtfoAvYuS1pEDtKngXQE1dcTDXR9dgwfqdakksFrhNHeKiqsYKD6KS5mga1NvegzbV6nKwsNyfGd",
```
is this really needed can't we just derive this from the mnemonic ?

Ashar2shahid commented on src/validator/validator.ts:
```diff
@@ -46,31 +57,134 @@ export const endpointSchema = z.object({
+
+export const logLevelSchema = z.union([z.literal('DEBUG'), z.literal('INFO'), z.literal('WARN'), z.literal('ERROR')]);
+
+export const nodeSettingsSchema = z.object({
```
Same here can't we use the airnode validator for this ?

acenolaza commented on src/validator/validator.ts:
```diff
@@ -1,8 +1,16 @@
-import { z, ZodFirstPartySchemaTypes } from 'zod';
+import { RefinementCtx, z, ZodFirstPartySchemaTypes } from 'zod';
+import { oisSchema } from './ois';
```
Tbh I would love to just re-use the airnode-validator package to avoid code duplication. Today I saw that Emanuel was applying a fix to the validator code and I thought to myself that I would need to also apply the same fix in Airkeeper 😓 

The thing is that functions in airnode-validator expect to receive an Airnode config.json file and I'm not sure what would happen if I try validating a file with extra fields. Maybe we need to expose functions in airnode-validator to validate parts of the config file like you are mentioning: `validateOis`, `validateNodeSettings`, etc.

I'd like to get @amarthadan opinion on this too.

Siegrift replied:
> Maybe we need to expose functions in airnode-validator to validate parts of the config file like you are mentioning: validateOis, validateNodeSettings, etc.

Ois schema is exported from validator and also used in https://github.com/api3dao/airnode/blob/master/packages/airnode-ois/src/index.ts#L11. Other schemas are exported as well, but they are not accessible that nicely. For example to import nodeSettingsSchema you have to do:

```
import {nodeSettingsSchema} from '@api3/airnode-validator/dist/src/config';
```

But this is a detail. I think you should be using the validations from Airnode.

amarthadan reviewed (changes requested):
First of all, sorry it took me so long to take a look :confused:

So, I think the main problem is in the configuration file you're proposing. I think that we don't actually need most of the stuff there. The point here is that most of the Airnode configuration file is not needed in Airkeeper and we want to add to the Airkeeper config just those fields from Airnode we actually need. For example, we don't need `nodeSettings`, do we? `triggers` (those Airnode specific)? Do we even need `ois`?. I think that if you get rid of everything we don't actually need for Airkeeper to work the whole thing will become much easier.

And yes, you shouldn't even write validation schemas for those parts that we need. Just include them from Airnode's validator. Take a look at how is that done in Airseeker, we're doing the same thing there.

acenolaza commented:
> First of all, sorry it took me so long to take a look confused

🙏🏻 

> So, I think the main problem is in the configuration file you're proposing. I think that we don't actually need most of the stuff there. The point here is that most of the Airnode configuration file is not needed in Airkeeper and we want to add to the Airkeeper config just those fields from Airnode we actually need. For example, we don't need `nodeSettings`, do we? `triggers` (those Airnode specific)? Do we even need `ois`?. I think that if you get rid of everything we don't actually need for Airkeeper to work the whole thing will become much easier.

At first I thought I could only take a few things from the Airnode config.json file and put it in airseeker.json file but the main issue with that is [here](https://github.com/api3dao/airkeeper/blob/d54168fd56f180e1ab2e5d9f4536136a4e1fd921/src/api/call-api.ts#L24) when we call the `airnode-node.callApi()` function that expects the whole config.json object as arg. I can probably hack it by sending only what is needed and casting the param object as `any` but I really wouldn't like to do that so instead I just copied everything from the config.json and put it airseeker.json. Basically `airnode-node.callApi()` requires `ois` and `apiCredentials` which we don't use in Airkeeper.

> And yes, you shouldn't even write validation schemas for those parts that we need. Just include them from Airnode's validator. Take a look at how is that done in Airseeker, we're doing the same thing there.

👍🏻

acenolaza commented:
What do you guys think of using airnode-validator package to parse and validate the Airnode related values in config.json and use the schema defined in validator.ts to parse and validate Airkeeper related values? https://github.com/api3dao/airkeeper/pull/62/commits/e08aa878857b2c309b856fa9541dae99f3e3bad2

amarthadan commented:
> What do you guys think of using `airnode-validator` package to parse and validate the Airnode related values in `config.json` and use the schema defined in `validator.ts` to parse and validate Airkeeper related values? [e08aa87](https://github.com/api3dao/airkeeper/commit/e08aa878857b2c309b856fa9541dae99f3e3bad2)

I think that would be the right approach but I'm wondering whether it makes sense to continue with this until we change `callApi` on Airnode side. I'm not sure it makes sense to merge the config if we need the whole thing. Maybe it would be worth investigating what changes would be needed on the Airnode side so we don't need the whole Airnode config. Or whether it's even possible.

acenolaza commented:
> > What do you guys think of using `airnode-validator` package to parse and validate the Airnode related values in `config.json` and use the schema defined in `validator.ts` to parse and validate Airkeeper related values? [e08aa87](https://github.com/api3dao/airkeeper/commit/e08aa878857b2c309b856fa9541dae99f3e3bad2)
> 
> I think that would be the right approach but I'm wondering whether it makes sense to continue with this until we change `callApi` on Airnode side. I'm not sure it makes sense to merge the config if we need the whole thing. Maybe it would be worth investigating what changes would be needed on the Airnode side so we don't need the whole Airnode config. Or whether it's even possible.

https://api3dao.atlassian.net/browse/AN-749

acenolaza commented:
@amarthadan I just finished the branch update but I also just realized that the [changes on callApi()](https://github.com/api3dao/airnode/blob/master/packages/airnode-node/src/api/index.ts#L270) to take only what's needed as arguments has not been released yet. Also, I think we'll need to export the api directory to be able to call `callApi()` from Airseeker. As an alternative we could call the [processHttpRequest() handler](https://github.com/api3dao/airnode/blob/master/packages/airnode-node/src/handlers/process-http-request.ts#L8) but that is still expecting the whole `Config` object.

acenolaza commented:
09/12 UPDATE:

1. I've updated all deps in package json. Airkeeper now supports `v0.8.0` packages. @vponline please review this PR since I've made many changes around gas price and logging since Airkeeper was also using `@api3/airnode-utilities` package.
2. Moved only required fields from Airnode's config.json into airkeeper.json. Only chains.options.fulfillmentGasLimit was left unused because `getGasPrice()` expects it as arg but this will be [addressed soon](https://github.com/api3dao/airnode/issues/1437).
3. Moved `airnodeAddress` and `airnodeXpub` into nodeSettings
4. Refactored `callApi()` to make `http-gateway` requests with only `ois` and `apiCredentials` as args (instead of the whole config object)
5. TODO: we need to add validation around the config schema (refinements) but I think that needs to be handled in a different issue.

vponline commented on test/config/config.ts:
```diff
@@ -14,60 +16,97 @@ export const buildAirnodeConfig = () => ({
-          unit: 'gwei',
-        },
+        fulfillmentGasLimit: 500000,
+        gasPriceOracle: [
```
We could also use `latestBlockPercentileGasPrice` and `providerRecommendedGasPrice` strategy configs here to make sure the E2E tests use them as well :thinking: 

It shouldn't affect the tests so just adding the strategies in the array should be fine, though

vponline commented on src/validator/validator.ts:
```diff
@@ -1,26 +1,31 @@
+export const rrpBeaconServerKeeperJobsTriggerSchema = z.object({
   chainIds: z.array(z.string()),
-  templateId: z.string(),
+  templateId: z.string().regex(/^0x[a-fA-F0-9]{64}$/),
```
Airnode exports an `evmIdSchema` that could work here?

acenolaza replied:
Yes! and I can probably also use the `evmAddressSchema` from airnode-validator

vponline reviewed (commented):
Found one more potential refactor possibility:

acenolaza commented on config/airkeeper.example.json:
```diff
@@ -1,15 +1,39 @@
+      "type": "evm",
+      "options": {
+        "fulfillmentGasLimit": 500000,
+        "gasPriceOracle": [
```
Good idea 👍🏻 I'll add it

amarthadan commented on src/validator/validator.ts:
```diff
@@ -1,26 +1,31 @@
   requestSponsor: z.string(),
 });
 
+// TODO: XOR?
```
XOR sounds like the thing we want, right?

acenolaza replied:
I never got an "official" answer. I just put that comment in there because I saw Airkeeper being used with only one kind of triggers at a time but I'm not 100% there isn't a case where we might need Airkeeper to update beacons using both trigger types. Or they might just need to be optional 🤷🏻 

@aquarat any thoughts?

aquarat replied:
I think the use of RRP for beacons is very unlikely in practice and they wouldn't be used by API3 in production. AFAIK they were a demonstration of the power of RRP as a tech stack component.

The only scenario I can think of where RRP beacons would be useful is where you have an API provider with an existing Airnode who is unlikely or unwilling to support PSP directly. Why would they not want to support PSP directly? Maybe they don't want their data _streamed_/made public/resold, in which case supporting it indirectly via RRP would probably displease them eventually. The idea is that Airkeeper and Airnode will eventually become one application, which will make RRP beacons even less likely.

Would there be a negative impact to supporting both protocols concurrently? Is it just a reserved concurrency question? It looks like Airkeeper doesn't care. If the impact is not negative then I'd go with OR (either or both). If there's the potential for complications down the line that could cost more time I'd make it XOR.

I think the main thing is that this is relatively unimportant, so feel free to pick one and go with it. 👍 🤷

amarthadan reviewed (approved):
Minor notes/questions, otherwise, LGTM :+1:

acenolaza commented on README.md:
```diff
@@ -29,14 +29,15 @@ updates is `12345` and for PSP it is `2`.
-  `config.json` and `secrets.env` file can be copied from the Airnode repo and placed in the /config directory of this
-  repo. Examples of these two files can be found in the /config directory of this repo.
+- Airkeeper configuration file will require you to copy over some parts of the Airnode's
+  [`config.json`](https://github.com/api3dao/airnode/blob/v0.8.0/packages/airnode-node/config/config.example.json) file.
```
It was because I think the user should be looking at the `config.json` file that matches the version of the airnode dependencies. Otherwise it might add or remove properties in airkeeper.json that are not supported by the airnode.

I know that it will be a hassle to always remember to update this link whenever we update the deps in package.json so should I maybe find a way to extract this value from there? 🤔

amarthadan replied:
I don't think it's that important at the moment but feel free to create an issue for that.

acenolaza commented on src/evm/initialize-provider.ts:
```diff
@@ -47,28 +48,10 @@ export const initializeEvmState = async (
   const currentBlockLog = utils.logger.pend('INFO', currentBlockMessage);
 
   // Fetch current gas fee data
-  const [gasPriceLogs, gasTarget] = await utils.getGasPrice({
-    provider,
-    chainOptions: chain.options,
-  });
-  if (!gasTarget) {
-    const message = 'Failed to fetch gas price';
-    const log = utils.logger.pend('ERROR', message);
-    return [[...gasPriceLogs, log], null];
-  }
-  let gasTargetMessage;
-  if (chain.options.txType === 'eip1559') {
-    const gweiMaxFee = node.evm.weiToGwei(gasTarget.maxFeePerGas!);
-    const gweiPriorityFee = node.evm.weiToGwei(gasTarget.maxPriorityFeePerGas!);
-    gasTargetMessage = `Gas price (EIP-1559) set to a Max Fee of ${gweiMaxFee} Gwei and a Priority Fee of ${gweiPriorityFee} Gwei`;
-  } else {
-    const gweiPrice = node.evm.weiToGwei(gasTarget.gasPrice!);
-    gasTargetMessage = `Gas price (legacy) set to ${gweiPrice} Gwei`;
-  }
-  const gasTargetLog = utils.logger.pend('INFO', gasTargetMessage);
```
I used [this PR](https://github.com/api3dao/airnode/pull/1321) for guidance. More specifically looked at changes in this file: packages/airnode-node/src/evm/handlers/process-transactions.ts

This was one of the reasons I asked @vponline to review this PR (the other being the refactor around logging)

vponline replied:
It won't be needed anymore as:
1. The gas price strategies will always return at least the constant strategy price as a last resort
2. EIP1559/Legacy logging are handled in the airnode-utilitites side

vponline commented on package.json:
```diff
@@ -52,49 +52,50 @@
+    "ethers": "^5.7.0",
     "lodash": "^4.17.21",
-    "zod": "^3.13.4"
+    "zod": "^3.19.0"
```
> Yes, I had issues with mixed zod releases in the past. I guess the version should then be [3.17.3](https://github.com/api3dao/airnode/blob/v0.8.0/packages/airnode-node/package.json#L42)
